### PR TITLE
[Merged by Bors] - feat: Consolidate authentication 

### PR DESCRIPTION
**Pull request by maltesander**
# Description

Changed authentication to list of authentication classes.

Multiple authentication providers can be used (file, ldap). Hot reloading works.

test: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/trino-operator-it-custom/70/

fixes: https://github.com/stackabletech/trino-operator/issues/401

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Review by sbernauer — CHANGES_REQUESTED** [review]
First batch of comments
Some todos for me:
comment on sidecar wegen schreiben in variable und nicht direkt ins file
comment zu logging warum immer angeschaltet und nicht merged config wird
resourcen für vector checken

File: rust/operator-binary/src/authentication/mod.rs [hunk 1]
```diff
@@ -0,0 +1,654 @@
...
+        // Properties like PASSWORD, CERTIFICATE are only added once and the order is important
+        // due to Trino starting to evaluate the authenticators depending on the given order
+        let mut http_server_authentication_types = vec![];
+
+        for auth_type in &trino_auth.authentication_types {
+            match auth_type {
+                TrinoAuthenticationType::Password(password_auth) => {
+                    if !http_server_authentication_types.contains(&auth_type.to_string()) {
+                        http_server_authentication_types.push(auth_type.to_string());
+                    }
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/authentication/mod.rs]: Good point! https://github.com/stackabletech/trino-operator/pull/434/commits/1616383f85f16bae589a0749dd1b73f3f9a7e853

File: rust/operator-binary/src/authentication/mod.rs [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+            .entry(role)
+            .or_insert(BTreeMap::new())
+            .entry(container)
+            .or_insert(Vec::new())
+            .extend(commands)
+    }
+
+    /// Add an additional volume for the pod builder.
+    pub fn add_volume(&mut self, volume: Volume) {
+        if !self.volumes.contains(&volume) {
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/authentication/mod.rs:189]: It derives PartialEq, basically checks the equality of all fields in that struct.

File: rust/operator-binary/src/authentication/mod.rs [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+
+        if !http_server_authentication_types.is_empty() {
+            authentication_config.add_config_property(
+                TrinoRole::Coordinator,
+                HTTP_SERVER_AUTHENTICATION_TYPE.to_string(),
+                http_server_authentication_types.join(","),
+            );
+        }
+
+        debug!(
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/authentication/mod.rs:102]: No strong opinion here, but trace is better probably.

File: docs/modules/trino/pages/usage_guide/security.adoc [hunk 4]
```diff
@@ -19,32 +21,46 @@ The https://trino.io/docs/current/security/password-file.html[file based authent
...
 [source,yaml]
 ----
 include::example$usage_guide/trino-file-auth-snippet.yaml[tag=trino]
 ----
 
+Multiple authentication classes with different user secrets can be provided.
+
+The operator never reads secrets directly, but mounts them directly into the `Pod`. Volume mount names can not exceed *63* characters due to Kubernetes restrictions.
+
+For uniqueness, the volume mount name is internally build up of the name of the `AuthenticationClass` and the name of the referenced `Secret` concatenated with a "-" character (e.g. "simple-trino-file-auth-simple-trino-users" for the example above). This means the `AuthenticationClass` and `Secret` names combined must be smaller than *62* characters.
```
[the remark this answers is not in the document]
Comment (maltesander) [reply docs/modules/trino/pages/usage_guide/security.adoc:44]: If i think about it yes, not sure why i used both auth class and secret name...ill test it without and adapt the docs.

File: docs/modules/trino/examples/usage_guide/trino-secure-internal-tls.yaml [hunk 5]
```diff
@@ -45,21 +42,30 @@ metadata:
...
         secret:
-          name: secret-provisioner-trino-internal-tls-ca
+          name: secret-provisioner-trino-internal-tls-ca # <2>
           namespace: default
         autoGenerate: true
 ---
+apiVersion: authentication.stackable.tech/v1alpha1
+kind: AuthenticationClass
+metadata:
+  name: trino-file-auth # <3>
```
[the remark this answers is not in the document]
Comment (maltesander) [reply docs/modules/trino/examples/usage_guide/trino-secure-internal-tls.yaml]: https://github.com/stackabletech/trino-operator/pull/434/commits/108862ed046ff40375b1e3825d81d687963e0d0e

File: rust/operator-binary/src/authentication/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/operator-binary/src/authentication/mod.rs:189]: Hm, in that case I would prefer checking for the same name

**Review by sbernauer — COMMENTED** [review]
Should be done wit Rust code

File: rust/operator-binary/src/authentication/password/mod.rs [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+use stackable_operator::{commons::product_image_selection::ResolvedProductImage, product_config};
+use stackable_trino_crd::{Container, TrinoRole, RW_CONFIG_DIR_NAME};
+use std::collections::BTreeMap;
+use tracing::debug;
+
+// Trino properties
+pub(crate) const PASSWORD_AUTHENTICATOR_CONFIG_FILES: &str = "password-authenticator.config-files";
+const PASSWORD_AUTHENTICATOR_NAME: &str = "password-authenticator.name";
+// file handling
+pub(crate) const CONFIG_FILE_NAME_SUFFIX: &str = ".properties";
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/authentication/password/mod.rs]: removed https://github.com/stackabletech/trino-operator/pull/434/commits/ac795ff149097f111e6b52be3fbe48ef37d986c5

File: docs/modules/trino/pages/usage_guide/security.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (maltesander) [reply docs/modules/trino/pages/usage_guide/security.adoc]: only authclass name now https://github.com/stackabletech/trino-operator/pull/434/commits/ac795ff149097f111e6b52be3fbe48ef37d986c5

File: rust/operator-binary/src/authentication/mod.rs [hunk 7]
```diff
@@ -0,0 +1,652 @@
...
+        source: product_config::writer::PropertiesWriterError,
+    },
+    #[snafu(display("Failed to configure trino password authentication"))]
+    InvalidPasswordAuthenticationConfig { source: password::Error },
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// This is the final product after iterating through all authenticators.
+/// Contains all relevant information about config files, volumes etc. to enable authentication.
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/authentication/mod.rs:51]: https://github.com/stackabletech/trino-operator/pull/434/commits/211d0b536b2d6cb8958f3d1f1fdba415d73e319d

File: rust/operator-binary/src/authentication/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/authentication/mod.rs]: https://github.com/stackabletech/trino-operator/pull/434/commits/15c81187481dd52e2db7ba0bd3402367d879a5b9

File: rust/operator-binary/src/authentication/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/authentication/mod.rs]: done https://github.com/stackabletech/trino-operator/pull/434/commits/d4ac56868cf6efb9bb82fe6777997bb8a036d56f

File: rust/operator-binary/src/authentication/password/file.rs [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+        stackable_auth_secret_dir = PASSWORD_AUTHENTICATOR_SECRET_MOUNT_PATH,
+    ));
+
+    cb_pw_file_updater
+        .image_from_product_image(resolved_product_image)
+        // calculated mounts
+        .add_volume_mounts(volume_mounts)
+        // fixed
+        .add_volume_mount("log", STACKABLE_LOG_DIR)
+        .resources(ResourceRequirements {
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/authentication/password/file.rs]: https://github.com/stackabletech/trino-operator/pull/434/commits/6ed1939c888e10c69266591a100ce7e30613b724

File: rust/operator-binary/src/authentication/password/file.rs [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+            requests: Some(resources.clone()),
+            limits: Some(resources),
+            ..ResourceRequirements::default()
+        })
+        .command(vec!["/bin/bash".to_string(), "-c".to_string()])
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/authentication/password/file.rs]: done works

File: rust/operator-binary/src/authentication/mod.rs [hunk 10]
```diff
@@ -0,0 +1,658 @@
...
+
+    /// Add an additional volume for the pod builder.
+    pub fn add_volume(&mut self, volume: Volume) {
+        if self
+            .volumes
+            .iter()
+            .filter(|r| r.name == volume.name)
+            .count()
+            == 0
+        {
```
Comment (sbernauer) [inline rust/operator-binary/src/authentication/mod.rs:200]:
```suggestion
        if !self.volumes.iter().any(|v| v.name == volume.name) {
```

File: rust/operator-binary/src/authentication/mod.rs [hunk 11]
```diff
@@ -0,0 +1,658 @@
...
+            .or_insert_with(BTreeMap::new)
+            .entry(container)
+            .or_insert_with(Vec::new);
+
+        if current_volume_mounts
+            .iter()
+            .filter(|r| r.name == volume_mount.name)
+            .count()
+            == 0
+        {
```
Comment (sbernauer) [inline rust/operator-binary/src/authentication/mod.rs:225]:
```suggestion
        if !current_volume_mounts
            .iter()
            .any(|vm| vm.name == volume_mount.name)
        {
```

File: rust/operator-binary/src/authentication/mod.rs [hunk 12]
```diff
@@ -0,0 +1,658 @@
...
+    /// Add an extra sidecar container for a given role
+    pub fn add_sidecar_container(&mut self, role: TrinoRole, container: Container) {
+        let containers_for_role = self.sidecar_containers.entry(role).or_insert_with(Vec::new);
+
+        if containers_for_role
+            .iter()
+            .filter(|r| r.name == container.name)
+            .count()
+            == 0
+        {
```
Comment (sbernauer) [inline rust/operator-binary/src/authentication/mod.rs:239]:
```suggestion
        if !containers_for_role.iter().any(|c| c.name == container.name) {
```

File: rust/operator-binary/src/authentication/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/operator-binary/src/authentication/mod.rs]: thx!

File: rust/operator-binary/src/controller.rs [hunk 13]
```diff
@@ -919,6 +923,7 @@ fn build_rolegroup_statefulset(
             "config",
             "log",
             merged_config.logging.containers.get(&Container::Vector),
+            ResourceRequirements::default(),
```
Comment (sbernauer) [inline rust/operator-binary/src/controller.rs:926]: We must pass in dedicated resources constructed using `ResourceRequirementsBuilder`.

**Review by sbernauer — CHANGES_REQUESTED** [review]

**Comment by sbernauer** [comment]
We have the same problem as in #433:
As we have decided to stuff everything into a single release we need to specify vector resources.
If you would be ok with it, I would push all the needed changes for the resources (including docs) to this branch, so that we merge them together.
Does that work for you?

**Comment by maltesander** [comment]
> We have the same problem as in #433: As we have decided to stuff everything into a single release we need to specify vector resources. If you would be ok with it, I would push all the needed changes for the resources (including docs) to this branch, so that we merge them together. Does that work for you?

Works for me thanks!

File: rust/operator-binary/src/authentication/password/file.rs [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+        .args(vec![commands.join(" && ")])
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/operator-binary/src/authentication/password/file.rs:178]: nevermind, script is newline separated and the && only for piping stdout/err

File: tests/templates/kuttl/smoke/00-limit-range.yaml [hunk 15]
```diff
@@ -0,0 +1,12 @@
+# FIXME: Blocked by e.g. zookeeper not having resources set
```
Comment (sbernauer) [inline tests/templates/kuttl/smoke/00-limit-range.yaml:1]: Marker

File: tests/templates/kuttl/smoke/helm-bitnami-minio-values.yaml.j2 [hunk 16]
```diff
@@ -1,4 +1,32 @@
...
+    value: "true"
+  - name: MINIO_LOG_LEVEL
+    value: DEBUG
+
+#defaultBuckets: hive
+
+provisioning:
+  enabled: true
+  buckets:
+    - name: hive
```
Comment (razvan) [inline tests/templates/kuttl/smoke/helm-bitnami-minio-values.yaml.j2:15]: For some reason I cannot make suggestions here, but please replace `hive` with `trino`

File: tests/templates/kuttl/smoke/04-prepare-bucket.yaml.j2 [hunk 17]
```diff
@@ -4,17 +4,11 @@ kind: TestStep
...
-{% else %}
 {% if test_scenario['values']['s3-use-tls'] == 'true' %}
-  - command: kubectl exec -n $NAMESPACE minio-client -- sh -c 'mc --insecure alias set minio https://minio:9000 trino trinotrino'
+  - command: kubectl exec -n $NAMESPACE minio-client -- sh -c 'mc --insecure alias set minio https://minio:9000 $$MINIO_SERVER_ACCESS_KEY $$MINIO_SERVER_SECRET_KEY'
 {% else %}
-  - command: kubectl exec -n $NAMESPACE minio-client -- sh -c 'mc --insecure alias set minio http://minio:9000 trino trinotrino'
-{% endif %}
+  - command: kubectl exec -n $NAMESPACE minio-client -- sh -c 'mc --insecure alias set minio http://minio:9000 $$MINIO_SERVER_ACCESS_KEY $$MINIO_SERVER_SECRET_KEY'
 {% endif %}
+  - command: kubectl exec -n $NAMESPACE minio-client -- mc mb minio/trino
```
Comment (razvan) [inline tests/templates/kuttl/smoke/04-prepare-bucket.yaml.j2:12]: This should already have been done by the `provisioner` of the Minio chart.

File: tests/templates/kuttl/smoke/helm-bitnami-minio-values.yaml.j2 (hunk 16 above)
Comment (razvan) [inline tests/templates/kuttl/smoke/helm-bitnami-minio-values.yaml.j2:15]:
```suggestion
    - name: trino
```

File: tests/templates/kuttl/smoke/00-secrets.yaml.j2 [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: centralized-minio-users
+type: Opaque
+stringData:
+  username1: |
+    username=hive
+    password=hivehive
```
Comment (razvan) [inline tests/templates/kuttl/smoke/00-secrets.yaml.j2:35]:
```suggestion
    username=trino
    password=trinotrino
```

File: tests/templates/kuttl/smoke/04-prepare-bucket.yaml.j2 (hunk 17 above)
Comment (razvan) [inline tests/templates/kuttl/smoke/04-prepare-bucket.yaml.j2:12]: The bucket is created by the chart provisioner.

```suggestion
```

**Review by razvan — COMMENTED** [review]
Only looked at the smoke/minio chart

**Comment by sbernauer** [comment]
@razvan my goal was to re-use the hive minio values so it was intentional. A better solution might be to call the bucket `test` everywhere. WDYT?

**Comment by sbernauer** [comment]
Started test: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/trino-operator-it-custom/71/

**Comment by sbernauer** [comment]
Next try: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/trino-operator-it-custom/72/

**Comment by sbernauer** [comment]
And another https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/trino-operator-it-custom/74/

**Comment by sbernauer** [comment]
https://logs.t2.stackable.tech/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-2h,to:now))&_a=(columns:!(message),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:t2-k8s-events,key:t2_cluster,negate:!f,params:(query:'2ed9a198-d2f8-45a4-abe5-ec6964c0c400'),type:phrase),query:(match_phrase:(t2_cluster:'2ed9a198-d2f8-45a4-abe5-ec6964c0c400'))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:t2-k8s-events,key:namespace,negate:!f,params:(query:kuttl-test-strong-hookworm),type:phrase),query:(match_phrase:(namespace:kuttl-test-strong-hookworm)))),index:t2-k8s-events,interval:auto,query:(language:kuery,query:''),sort:!())
`create Pod trino-coordinator-default-0 in StatefulSet trino-coordinator-default failed error: Pod "trino-coordinator-default-0" is invalid: [spec.volumes[4].name: Invalid value: "trino-with-ldap-bind-kuttl-test-strong-hookworm-bind-credentials": must be no more than 63 characters, spec.containers[1].volumeMounts[1].name: Not found: "trino-with-ldap-bind-kuttl-test-strong-hookworm-bind-credentials", spec.initContainers[0].volumeMounts[0].name: Not found: "trino-with-ldap-bind-kuttl-test-strong-hookworm-bind-credentials"]`

**Comment by sbernauer** [comment]
Started test https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/trino-operator-it-custom/75/

File: rust/crd/Cargo.toml [hunk 19]
```diff
@@ -9,7 +9,7 @@ version = "0.0.0-dev"
 publish = false
 
 [dependencies]
-stackable-operator = { git = "https://github.com/stackabletech/operator-rs.git", tag = "0.41.0" }
+stackable-operator = { git = "https://github.com/stackabletech/operator-rs.git", tag = "0.42.1" }
```
Comment (maltesander) [inline rust/crd/Cargo.toml:12]: Lets go for 0.42.2 for smaller CRDS!

File: rust/operator-binary/Cargo.toml [hunk 20]
```diff
@@ -10,12 +10,12 @@ version = "0.0.0-dev"
 publish = false
 
 [dependencies]
-stackable-operator = { git = "https://github.com/stackabletech/operator-rs.git", tag = "0.41.0" }
+stackable-operator = { git = "https://github.com/stackabletech/operator-rs.git", tag = "0.42.1" }
```
Comment (maltesander) [inline rust/operator-binary/Cargo.toml:13]: bump operator-rs

**Comment by sbernauer** [comment]
Hopefully last test run: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/trino-operator-it-custom/76/

**Comment by sbernauer** [comment]
The one and only https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/trino-operator-it-custom/77/

**Comment by sbernauer** [comment]
bors r+
Tests passed :rocket:

**Review by sbernauer — APPROVED** [review]
Approving together with Malte, as we have both written parts of this PR

**Pull request closed without merging**